### PR TITLE
refactor: Poll Context object for API

### PR DESCRIPTION
**Pull request by astubbs**
Provides:

- an organised view per TopicPartition
- failure counts for records - which will improve the retry system
- Enables all api's to be batched - not just special batch methods
- allow for future extensions without breaking primary APIs

This will be a minor breaking change

- [x] Replace work container look up through work manager with context object reference
- [x] Upgrade docs (breaking change)
- [x] Update docs (features and batching section)
- [x] Implement collection
- [x] Full failure history, not just counts - memory leak potential - cap at something reasonable (customisable)
- [x] Unify batch and non batch flows
- [x] Update retry provider
- [x] Hide work container access from public api through protected inheritance from internal package
- [x] Remove failure history to separate PR
- [x] Add record id(tp, offset) - useful for user to identify records. 
- [x] For hiding work container, try out delegate instead of inheritance

### CODE REVIEW

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/PollContext.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+
+    /**
+     * Throws {@link IllegalStateException} if a {@link ParallelConsumerOptions#getBatchSize()} has been set.
+     *
+     * @return the single record entry in this context
+     */
+    public RecordContext<K, V> getSingleRecord() {
+        var recordList = getContextsFlattened();
+        if (recordList.size() != 1) {
+            // todo docs - same message as before
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/PollContext.java:63]: docs

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/RecordContext.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+import io.confluent.parallelconsumer.state.WorkContainer;
+import lombok.*;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+
+@AllArgsConstructor
+@Builder(toBuilder = true)
+@Value
+public class RecordContext<K, V> {
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/RecordContext.java:14]: docs

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/PollContext.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+import java.util.*;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+@AllArgsConstructor
+@Value(staticConstructor = "of")
+@Getter(AccessLevel.NONE)
+//@Setter(AccessLevel.NONE)
+public class PollContext<K, V> implements Iterable<RecordContext<K, V>> {
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/PollContext.java:24]: docs

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java [hunk 4]
```diff
@@ -125,22 +99,22 @@ public void pollAndProduce(Function<ConsumerRecord<K, V>, ProducerRecord<K, V>>
...
-
-        Function<List<ConsumerRecord<K, V>>, List<Object>> wrappedUserFunc = (recordList) -> {
-            log.trace("asyncPoll - Consumed set of records ({}), executing void function...", recordList.size());
-            usersVoidConsumptionFunction.accept(recordList);
-            return UniLists.of(); // user function returns no produce records, so we satisfy our api
-        };
-        Consumer<Object> voidCallBack = (ignore) -> log.trace("Void callback applied.");
-        supervisorLoop(wrappedUserFunc, voidCallBack);
-    }
+//
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelEoSStreamProcessor.java:106]: delete

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelStreamProcessor.java [hunk 5]
```diff
@@ -41,47 +40,43 @@
...
+//     * The system will treat the messages as a set, so if an error is thrown by the user code, then all messages will be
+//     * marked as failed and be retried (Note that when they are retried, there is no guarantee they will all be in the
+//     * same batch again). So if you're going to process messages individually, then don't use this function.
+//     * <p>
+//     * Otherwise, if you're going to process messages in sub sets from this batch, it's better to instead adjust the
+//     * {@link ParallelConsumerOptions#getBatchSize()} instead to the actual desired size, and process them as a whole.
+//     *
+//     * @see ParallelConsumerOptions#getBatchSize()
+//     */
+//    void pollBatch(Consumer<PollContext<K, V>> usersVoidConsumptionFunction);
```
Comment (astubbs) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/ParallelStreamProcessor.java:79]: delete

File: parallel-consumer-reactor/src/main/java/io/confluent/parallelconsumer/reactor/ReactorProcessor.java [hunk 6]
```diff
@@ -66,51 +64,51 @@ public void close(Duration timeout, DrainingMode drainMode) {
...
+//     * Make sure that you do any work immediately in a Publisher / Flux - do not block this thread.
+//     * <p>
+//     * The system will treat the messages as a set, so if an error is thrown by the user code, then all messages will be
+//     * marked as failed and be retried (Note that when they are retried, there is no guarantee they will all be in the
+//     * same batch again). So if you're going to process messages individually, then don't use this function.
+//     * <p>
+//     * Otherwise, if you're going to process messages in sub sets from this batch, it's better to instead adjust the
+//     * {@link ParallelConsumerOptions#getBatchSize()} instead to the actual desired size, and process them as a whole.
+//     *
+//     * @param reactorFunction user function that takes a single record, and returns some type of Publisher to process
```
Comment (astubbs) [inline parallel-consumer-reactor/src/main/java/io/confluent/parallelconsumer/reactor/ReactorProcessor.java:81]: delete

File: parallel-consumer-reactor/src/main/java/io/confluent/parallelconsumer/reactor/ReactorProcessor.java [hunk 7]
```diff
@@ -161,33 +146,29 @@ public void reactBatch(Function<List<ConsumerRecord<K, V>>, Publisher<?>> reacto
...
-            return publisher;
-        };
-
-        //
-        reactBatch(batchReactorFunctionWrapper);
-    }
+//    /**
+//     * Make sure that you do any work immediately in a Publisher / Flux - do not block this thread.
+//     * <p>
+//     * Like {@link #reactBatch} but no batching - single message at a time.
```
Comment (astubbs) [inline parallel-consumer-reactor/src/main/java/io/confluent/parallelconsumer/reactor/ReactorProcessor.java:152]: delete

File: parallel-consumer-vertx/src/main/java/io/confluent/parallelconsumer/vertx/VertxParallelEoSStreamProcessor.java [hunk 8]
```diff
@@ -185,15 +182,13 @@ public void vertxHttpWebClient(BiFunction<WebClient, ConsumerRecord<K, V>, Futur
         super.supervisorLoop(userFuncWrapper, noOp);
     }
 
-    private void addVertxHooks(ConsumerRecord<K, V> consumerRecord, Future<?> send) {
-        addVertxHooks(UniLists.of(consumerRecord), send);
-    }
+//    private void addVertxHooks(PollContext<K, V> consumerRecord, Future<?> send) {
```
Comment (astubbs) [inline parallel-consumer-vertx/src/main/java/io/confluent/parallelconsumer/vertx/VertxParallelEoSStreamProcessor.java:185]: delete

**Review by astubbs — COMMENTED** [review]
todos

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/PollContext.java [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+    /**
+     * For backwards compatiabilty with {@link ConsumerRecord#value()}
+     */
+    public V value() {
+        return getSingleConsumerRecord().value();
+    }
+
+    /**
+     * For backwards compatiabilty with {@link ConsumerRecord#value()}
```
Comment (nioertel) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/PollContext.java:80]: copy/paste: `...#key()`

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/PollContext.java [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+     */
+    public K key() {
+        return getSingleConsumerRecord().key();
+    }
+
+    /**
+     * For backwards compatiabilty with {@link ConsumerRecord#value()}
```
Comment (nioertel) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/PollContext.java:87]: copy/paste: `...#offset()`

**Comment by astubbs** [comment]
Thanks for the review @nioertel !

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/PollContext.java [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+import java.util.stream.Stream;
+
+/**
+ * TODO docs
+ */
+@AllArgsConstructor
+@Value(staticConstructor = "of")
+@Getter(AccessLevel.NONE)
+//@Setter(AccessLevel.NONE)
+public class PollContext<K, V> implements Iterable<RecordContext<K, V>> {
```
[the remark this answers is not in the document]
Comment (astubbs) [reply parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/PollContext.java:43]: @nioertel ? @JorgenRingen ?

**Comment by astubbs** [comment]
PR feature complete

**Review by astubbs — COMMENTED** [review]
lgtm

**Review by astubbs — COMMENTED** [review]
lgtm

File: parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/JStreamParallelStreamProcessor.java [hunk 12]
```diff
@@ -14,7 +14,7 @@
 
 public interface JStreamParallelStreamProcessor<K, V> extends DrainingCloseable {
 
-    static <KK, VV> JStreamParallelStreamProcessor<KK, VV> createJStreamEosStreamProcessor(ParallelConsumerOptions options) {
+    static JStreamParallelStreamProcessor createJStreamEosStreamProcessor(ParallelConsumerOptions<?, ?> options) {
```
Comment (acktsap) [inline parallel-consumer-core/src/main/java/io/confluent/parallelconsumer/JStreamParallelStreamProcessor.java:17]: Why generic is removed in the factory method?

**Pull request merged**